core/src/avm1/value.rs

```diff
@@ -483,6 +481,223 @@ impl<'gc> Value<'gc> {
+                    buf.push(digit());
+                }
+                exp = 0;
+            }
+            _ => {
+                // 1.345e-15
+                buf.push(b'0');
+                let n = digit();
+                if n != 0 {
+                    buf.push(n);
+                }
+                buf.push(b'.');
+                for _ in 0..MAX_DECIMAL_PLACES - 1 {
+                    buf.push(digit());
+                }
+            }
+        };
+
+        // Rounding: Peek at the next generated digit and round accordingly.
+        // Ties round away from zero.
+        if digit() >= b'5' {
+            // Add 1 to the right-most digit, carrying if we hit a 9.
+            for c in buf.iter_mut().rev() {
+                if *c == b'9' {
+                    *c = b'0';
+                } else if *c >= b'0' {
+                    *c += 1;
+                    break;
+                }
+            }
+        }
+
+        // Trim any trailing zeros and decimal point.
+        while buf.last() == Some(&b'0') {
+            buf.pop();
+        }
+        if buf.last() == Some(&b'.') {
+            buf.pop();
+        }
+
+        let mut start = 0;
+        if exp != 0 {
+            // Write exponent (e+###).
+
+            // Lots of band-aids here to attempt to clean up the rounding above.
+            // None of this accounts for the `-` in negative values, causing quite a few bugs.
+            // PLAYER-SPECIFIC: I think these checks were added in Flash Player 6.
+            // Trim leading zeros.
+            let pos = buf.iter().position(|&c| c != b'0').unwrap_or(buf.len());
+            if pos != 0 {
+                buf.copy_within(pos.., 0);
+                buf.truncate(buf.len() - pos);
+            }
+            if buf.is_empty() {
+                // Fix up 9.99999 being rounded to 0.00000 when there is no space for the carried 1.
+                // If we have no digits, the value was all 0s that were trimmed, so round to 1.
+                buf.push(b'1');
+                exp += 1;
+            } else {
+                // Fix up 100e15 to 1e17.
+                let pos = buf.iter().rposition(|&c| c != b'0').unwrap_or_default();
+                if pos == 0 {
+                    exp += buf.len() as i32 - 1;
+                    buf.truncate(1);
+                }
+            }
+            let _ = write!(&mut buf, "e{:+}", exp);
+        }
+
+        // One final band-aid to eliminate any leading zeros.
+        let i = if is_negative { 1 } else { 0 };
+        if buf.get(i) == Some(&b'0') && buf.get(i + 1) != Some(&b'.') {
+            if i > 0 {
+                buf[i] = buf[i - 1];
+                start = i;
+            } else {
+                start = i + 1;
+            }
+        }
+
+        // SAFETY: Buffer is guaranteed to only contain ASCII digits.
+        let s = unsafe { std::str::from_utf8_unchecked(&buf[start..]) };
+        s.to_string().into()
```

Comment:
can we turn this straight into AvmString, without going through a temporary String copy?

Reply:
I believe this ends up reusing the allocation in `AvmString::new_utf8`, so hopefully shouldn't copy. It could avoid another UTF-8/ASCII check if we added an `unsafe fn AvmString::new_utf8_unchecked` or similar.

Reply:
Oh, my bad, I thought the method was returning Wstr/AvmString already.